File: services/sc-api-gateway/app/schemas/matches.py
from datetime import datetime
from pydantic import BaseModel, field_validator, model_validator


class MatchCreateResponse(BaseModel):
    match_id: str
    status: str


class RoiPoint(BaseModel):
    x: float
    y: float


class MatchConfigRequest(BaseModel):
    roi_polygon: list[RoiPoint]
    start_seconds: float
    end_seconds: float

    @field_validator('roi_polygon')
    @classmethod
    def min_3_points(cls, v):
        if len(v) < 3:
            raise ValueError('roi_polygon mínim 3 punts')
        return v

    @model_validator(mode='after')
    def end_after_start(self):
        if self.end_seconds <= self.start_seconds:
            raise ValueError('end_seconds ha de ser major que start_seconds')
        return self


class MatchConfigResponse(BaseModel):
    match_id: str
    roi_polygon: list[RoiPoint]
    start_seconds: float
    end_seconds: float


class MatchListItem(BaseModel):
    match_id: str
    title: str
    status: str
    created_at: datetime
    start_seconds: float | None = None
    end_seconds: float | None = None
    has_roi: bool


class ProcessMatchResponse(BaseModel):
    match_id: str
    status: str


class MatchDetail(BaseModel):
    match_id: str
    title: str
    status: str
    created_at: datetime
    start_seconds: float | None = None
    end_seconds: float | None = None
    download_url: str | None = None
    ai_stats: dict | None = None
    ai_report: str | None = None

File: services/sc-frontend/src/api/matches.ts
import apiClient from './client';

export interface RoiPoint {
  x: number;
  y: number;
}

interface MatchCreateResponse {
  match_id: string;
  status: string;
}

interface MatchConfigResponse {
  match_id: string;
  roi_polygon: RoiPoint[];
  start_seconds: number;
  end_seconds: number;
}

export interface MatchListItem {
  match_id: string;
  title: string;
  status: string;
  created_at: string;
  start_seconds: number | null;
  end_seconds: number | null;
  has_roi: boolean;
}

export async function getMatches(): Promise<MatchListItem[]> {
  const { data } = await apiClient.get<MatchListItem[]>('/api/v1/matches');
  return data;
}

export async function createMatch(
  video: File,
  title: string,
  onProgress?: (pct: number) => void,
): Promise<MatchCreateResponse> {
  const form = new FormData();
  form.append('video', video);
  form.append('title', title);
  const { data } = await apiClient.post<MatchCreateResponse>('/api/v1/matches', form, {
    onUploadProgress(e) {
      if (e.total && onProgress) onProgress(Math.round((e.loaded / e.total) * 100));
    },
  });
  return data;
}

export async function updateMatchConfig(
  matchId: string,
  roiPolygon: RoiPoint[],
  startSeconds: number,
  endSeconds: number,
): Promise<MatchConfigResponse> {
  const { data } = await apiClient.patch<MatchConfigResponse>(
    `/api/v1/matches/${matchId}/config`,
    { roi_polygon: roiPolygon, start_seconds: startSeconds, end_seconds: endSeconds },
  );
  return data;
}

export async function processMatch(matchId: string): Promise<{ match_id: string; status: string }> {
  const { data } = await apiClient.post<{ match_id: string; status: string }>(
    `/api/v1/matches/${matchId}/process`,
  );
  return data;
}

export interface AiStats {
  total_frames: number;
  avg_players_per_frame: number;
  avg_own_per_frame: number;
  avg_other_per_frame: number;
  pct_own: number;
  avg_confidence: number;
  max_density_frame: number;
  max_density_time_s: number;
  max_density_count: number;
  duration_s: number;
}

export interface MatchDetail {
  match_id: string;
  title: string;
  status: string;
  created_at: string;
  start_seconds: number | null;
  end_seconds: number | null;
  download_url: string | null;
  ai_stats: AiStats | null;
  ai_report: string | null;
  ai_report_refined: string | null;
}

export async function getMatch(matchId: string): Promise<MatchDetail> {
  const { data } = await apiClient.get<MatchDetail>(`/api/v1/matches/${matchId}`);
  return data;
}

export async function deleteMatch(matchId: string): Promise<void> {
  await apiClient.delete(`/api/v1/matches/${matchId}`);
}

export async function refineAiReport(matchId: string, userContext: string): Promise<string> {
  const { data } = await apiClient.post<{ ai_report: string }>(
    `/api/v1/matches/${matchId}/ai-refine`,
    { user_context: userContext },
  );
  return data.ai_report;
}
